storage/innobase/log/log0recv.cc

```diff
@@ -4841,6 +4843,50 @@ bool recv_dblwr_t::validate_page(const page_id_t page_id, lsn_t max_lsn,
   goto check_if_corrupted;
 }
 
+bool recv_dblwr_t::find_encrypted_page(const fil_node_t &node,
+                                       uint32_t page_no,
+                                       byte *buf)
+{
+  ut_ad(node.space->crypt_data);
+  ut_ad(node.space->full_crc32());
+  mysql_mutex_lock(&recv_sys.mutex);
+  const lsn_t max_lsn{log_sys.get_lsn()};
+  for (list::iterator page_it= pages.begin(); page_it != pages.end();
+       page_it++)
+  {
+    const byte *page= *page_it;
+    if (page_get_page_no(page) != page_no)
+      continue;
+    const lsn_t lsn= mach_read_from_8(page + FIL_PAGE_LSN);
+    if (buf_page_is_corrupted(lsn <= max_lsn, page, node.space->flags))
+      continue;
```

Comment:
If `lsn` is in the future, we would seem to wrongly approve the copy of the page in the doublewrite buffer. We must only approve pages that are between the latest checkpoint LSN and the end of the log. This filtering can actually be done already in `buf_dblwr_t::recover()`.